### PR TITLE
feat(wmg): gene info

### PR DESCRIPTION
**Pull request by atarashansky**
## Reason for Change

- #3723 

## Changes

- added gene info feature
- removed right sidebar collapse behavior. closing the right sidebar removes it entirely now
- added gene info buttons to x axis and genes in marker gene panel
- clicking gene info in marker gene panel opens bottom gene info drawer

## Testing steps

- [Rdev stack here](https://gene-info-frontend.rdev.single-cell.czi.technology/gene-expression)
- Gene info e2e tests added
- Tested locally
    - e2e tests all pass locally
    - Opening/closing behavior matches what is expected in ticket
    - Styles match Figma

## Demo

https://user-images.githubusercontent.com/109984998/227580723-76d9014a-0c7c-49ed-baf4-d04ab4c04bcc.mov

## Notes for Reviewer
@tihuan for readability
@atarashansky and @ashin-czi for functionality

### CODE REVIEW

File: frontend/src/common/queries/wheresMyGene.ts [hunk 1]
```diff
@@ -1139,11 +1146,37 @@ export async function fetchMarkerGenes({
...
+
+export async function fetchGeneInfo({
+  geneID,
+  geneSymbol,
+  signal,
+}: FetchGeneInfoParams): Promise<GeneInfo> {
+  const url =
+    API_URL + API.WMG_GENE_INFO + `?geneID=${geneID}&gene=${geneSymbol}`;
+  const response = await fetch(url, {
+    ...DEFAULT_FETCH_OPTIONS,
```
Comment (ashin-czi) [inline frontend/src/common/queries/wheresMyGene.ts:1158]: needed for CORS

File: frontend/src/components/common/SideBar/style.ts [hunk 2]
```diff
@@ -17,10 +17,12 @@ interface PositionerProps {
   isExpanded: boolean;
 }
 
+export const SIDEBAR_BOX_SHADOW_COLOR = "rgba(16, 22, 26, 0.15)";
```
Comment (ashin-czi) [inline frontend/src/components/common/SideBar/style.ts:20]: this color will be used as a separator line for the split right sidebar, top and bottom views

File: frontend/src/views/WheresMyGene/components/CellInfoSideBar/index.tsx [hunk 3]
```diff
@@ -241,7 +248,26 @@ function CellInfoSideBar({
...
+                  <InfoButtonWrapper
+                    onClick={() => {
+                      generateGeneInfo(gene[0]);
+
+                      track(EVENTS.WMG_FMG_GENE_INFO, {
+                        gene: gene[0],
+                      });
+                    }}
+                  >
+                    <StyledImage
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/CellInfoSideBar/index.tsx:262]: Adding icon button next to genes in FMG panel

File: frontend/src/views/WheresMyGene/components/GeneInfoSideBar/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneInfoSideBar/index.tsx:1]: The component for the gene info view

calls the new useGeneInfo hook to retrieve the gene summaries and other fields from the BE

File: frontend/src/views/WheresMyGene/components/GeneInfoSideBar/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneInfoSideBar/style.ts:1]: Just stylings for the gene info component

File: frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx [hunk 4]
```diff
@@ -18,19 +21,25 @@ import {
...
 
 function GeneButton({
   geneName,
   genesToDelete,
+  generateGeneInfo,
 }: {
   geneName: string;
   genesToDelete: string[];
   handleGeneClick: (gene: string) => void;
+  generateGeneInfo: (gene: string) => void;
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx:42]: passing in generateGeneInfo function to generate gene info when icon button is clicked on the x axis

File: frontend/src/views/WheresMyGene/components/HeatMap/components/YAxisChart/style.ts [hunk 5]
```diff
@@ -77,7 +77,7 @@ export const FlexRow = styled.div`
 `;
 
 export const InfoButtonWrapper = styled.div`
-  padding-left: 2px;
+  padding-left: 4px;
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/YAxisChart/style.ts:80]: increasing info button padding next to cell type names on y axis

File: frontend/src/views/WheresMyGene/components/Main/index.tsx [hunk 6]
```diff
@@ -271,27 +295,34 @@ export default function WheresMyGene(): JSX.Element {
...
-          testId="cell-type-details-panel"
-          disabled={false}
-          forceOpen={forceOpen}
-          wmgSideBar
-          width={CELL_INFO_SIDEBAR_WIDTH_PX}
-          truncatedLabel={`${tissuesByID[cellInfoCellType.tissueID].name} - ${
-            cellInfoCellType.cellType.name
-          }`}
-        >
+      {cellInfoCellType && tissuesByID ? (
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Main/index.tsx:298]: if cell info is populated then open up the right sidebar for FMG

File: frontend/src/views/WheresMyGene/components/Main/index.tsx [hunk 7]
```diff
@@ -271,27 +295,34 @@ export default function WheresMyGene(): JSX.Element {
...
           <CellInfoBar
+            generateGeneInfo={generateGeneInfo}
             cellInfoCellType={cellInfoCellType}
             tissueInfo={tissuesByID[cellInfoCellType.tissueID]}
+            handleClose={handleCloseRightSideBar}
+            title={`${cellInfoCellType.cellType.name}`}
           />
-        </SideBar>
+
+          {geneInfoGene && (
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Main/index.tsx:308]: if FMG panel is open AND gene info is populated, then split up right sidebar

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:1]: New component for right sidebar

Takes in children component(s). If there's only one component passed in then the full right sidebar will be rendered. If two components are passed in then the right sidebar will be split into upper and lower views

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+const FULL_MAX_HEIGHT = 100;
+const UPPER_SECTION_MAX_HEIGHT = 60;
+const DRAWER_MAX_HEIGHT = FULL_MAX_HEIGHT - UPPER_SECTION_MAX_HEIGHT;
+
+export default memo(function RightSideBar({
+  children,
+  width = CELL_INFO_SIDEBAR_WIDTH_PX,
+  testId,
+}: Props): JSX.Element | null {
+  const content = Children.map(
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:30]: getting all child components as an array

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    children,
+    (child) => child as ReactElement<RightSidebarProperties>
+  );
+
+  if (!content) return null;
+
+  const isSplit = content.length === 2;
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:37]: we determine if the right sidebar should be split if there are two children components

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            sdsIcon="xMark"
+            sdsSize="medium"
+            onClick={() => content[0].props.handleClose()}
+            sdsType="tertiary"
+          />
+        </HeaderContainer>
+        {content[0]}
+      </RightSideBarPositioner>
+
+      {isSplit && (
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:61]: split the component into two if there are two components

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+  return (
+    <SideBarWrapper
+      sideBarWidth={width}
+      position={Position.RIGHT}
+      data-test-id={testId}
+    >
+      <RightSideBarPositioner
+        isExpanded={true}
+        maxHeight={isSplit ? UPPER_SECTION_MAX_HEIGHT : FULL_MAX_HEIGHT}
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:47]: we set maxHeight for [styled.ts](https://github.com/chanzuckerberg/single-cell-data-portal/pull/4364/files#diff-d991f48498c220fb15e37496e7a5858888d6ece6c49b2a60df9f2aeb276c7504R28) to calculate the heights of the  components in the split sidebar

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+      >
+        <HeaderContainer>
+          <StyledTitle>{content[0].props.title}</StyledTitle>
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:50]: safe to index as `content[0]` here because the RightSideBar component requires at least one child

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+            sdsSize="medium"
+            onClick={() => content[0].props.handleClose()}
+            sdsType="tertiary"
+          />
+        </HeaderContainer>
+        {content[0]}
+      </RightSideBarPositioner>
+
+      {isSplit && (
+        <RightSideBarPositioner isExpanded={true} maxHeight={DRAWER_MAX_HEIGHT}>
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:62]: set the height for the bottom "drawer" view

File: frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+}
+interface Props {
+  children: ReactNode;
+  width?: number;
+  testId?: string;
+}
+
+// Values are in CSS vh units
+const FULL_MAX_HEIGHT = 100;
+const UPPER_SECTION_MAX_HEIGHT = 60;
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/index.tsx:22]: 60% of the viewport height for the FMG panel when in split view

File: frontend/src/views/WheresMyGene/components/RightSideBar/style.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+export const StyledTitle = styled.div`
+  font-weight: 600;
+  font-size: 20px;
+  line-height: 24px;
+`;
+
+/**
+ * Max height for the right sidebar should be 100vh - (header height)
+ */
+function sectionMaxheight({ maxHeight }: { maxHeight: number }) {
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/style.ts:28]: these are passed as props for the styled components

File: frontend/src/views/WheresMyGene/components/RightSideBar/style.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import styled from "@emotion/styled";
+import {
+  SideBarPositioner as RawSideBarPositioner,
+  SIDEBAR_BOX_SHADOW_COLOR,
+} from "src/components/common/SideBar/style";
+import { HEADER_HEIGHT_PX } from "src/components/Header/style";
+
+export const RightSideBarPositioner = styled(RawSideBarPositioner)`
+  max-height: ${sectionMaxheight};
+  box-shadow: inset 0px 1px 0px ${SIDEBAR_BOX_SHADOW_COLOR};
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/style.ts:10]: separator line between the two components for split view

File: frontend/src/views/WheresMyGene/components/RightSideBar/style.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import styled from "@emotion/styled";
+import {
+  SideBarPositioner as RawSideBarPositioner,
+  SIDEBAR_BOX_SHADOW_COLOR,
+} from "src/components/common/SideBar/style";
+import { HEADER_HEIGHT_PX } from "src/components/Header/style";
+
+export const RightSideBarPositioner = styled(RawSideBarPositioner)`
+  max-height: ${sectionMaxheight};
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/style.ts:9]: calculating the max height of the right sidebar components for single and split cases

File: frontend/src/views/WheresMyGene/components/RightSideBar/style.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+  font-size: 20px;
+  line-height: 24px;
+`;
+
+/**
+ * Max height for the right sidebar should be 100vh - (header height)
+ */
+function sectionMaxheight({ maxHeight }: { maxHeight: number }) {
+  return `calc(${maxHeight}vh - ${
+    maxHeight === 100 ? HEADER_HEIGHT_PX : HEADER_HEIGHT_PX / 2
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/RightSideBar/style.ts:30]: we are dividing the header height by two for the split right sidebar so both components equal to HEADER_HEIGHT_PX (48px) so that the full length of the right sidebar does not exceed the viewport

File: frontend/src/common/API.ts
Comment (ashin-czi) [inline frontend/src/common/API.ts:1]: not important

File: frontend/src/common/analytics/events.ts
Comment (ashin-czi) [inline frontend/src/common/analytics/events.ts:1]: not important

File: frontend/src/common/queries/entities.ts
Comment (ashin-czi) [inline frontend/src/common/queries/entities.ts:1]: not important

File: frontend/src/common/queries/wheresMyGene.ts
Comment (ashin-czi) [inline frontend/src/common/queries/wheresMyGene.ts:1]: Changes to call BE for gene info data

File: frontend/src/components/common/SideBar/style.ts [hunk 19]
```diff
@@ -105,18 +107,3 @@ export const SideBarOpenButtonWrapper = styled(SideBarToggleButtonWrapper)`
     width: 100%;
   }
 `;
-
-export const GeneSideBarOpenButtonWrapper = styled(SideBarToggleButtonWrapper)`
```
Comment (ashin-czi) [inline frontend/src/components/common/SideBar/style.ts:109]: not used anymore with introduction of new RightSideBar component

File: frontend/src/views/WheresMyGene/common/store/reducer.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/common/store/reducer.ts:1]: Changes for new hook to retrieve gene info data from BE

File: frontend/src/views/WheresMyGene/components/GeneInfoSideBar/style.ts
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/GeneInfoSideBar/style.ts:1]: Just stylings for gene info component

File: frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx [hunk 20]
```diff
@@ -66,14 +75,41 @@ function GeneButton({
...
+              generateGeneInfo(geneName);
+
+              track(EVENTS.WMG_GENE_INFO, {
+                gene: geneName,
+              });
+
+              if (!dispatch) return;
+
+              // Clear cell type info here so that FMG panel closes
+              dispatch(clearCellInfoCellType());
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx:90]: note that we clear cell info types here because when the gene info button is clicked on the xaxis then the FMG panel should also be cleared and replaced with the gene info panel

File: frontend/src/views/WheresMyGene/components/Main/index.tsx [hunk 21]
```diff
@@ -271,27 +295,34 @@ export default function WheresMyGene(): JSX.Element {
...
+            <GeneInfoBar
+              geneInfoGene={geneInfoGene}
+              handleClose={handleCloseGeneInfoSideBar}
+              title={`${geneInfoGene}`}
+            />
+          )}
+        </RightSideBar>
+      ) : (
+        geneInfoGene && (
```
Comment (ashin-czi) [inline frontend/src/views/WheresMyGene/components/Main/index.tsx:317]: if only gene info is populated then render the gene info panel at full length (not split)

**Comment by ashin-czi** [comment]
converting back to draft because e2e tests need to be made

File: frontend/src/common/queries/wheresMyGene.ts (hunk 1 above)
Comment (dsadgat) [reply frontend/src/common/queries/wheresMyGene.ts:1158]: this might be useful information in the code

**Comment by tihuan** [comment]
@signechambers1 do we care about how many people click on the gene info NCBI link? Right now I don't think we capture Plausible event for that!

**Comment by tihuan** [comment]
@ashin-czi currently `WMG_FMG_GENE_INFO` only sends the gene name, do we know if @signechambers1 would want the associated tissue in the payload too?

Right now `WMG_FMG_INFO_CLICKED` does have the tissue info in it, so technically Amanda could infer the tissue `WMG_FMG_GENE_INFO` is associated with by finding the last `WMG_FMG_INFO_CLICKED`. But not sure if that's too much work for them 😆

File: frontend/src/views/WheresMyGene/components/CellInfoSideBar/index.tsx [hunk 22]
```diff
@@ -241,7 +248,27 @@ function CellInfoSideBar({
           <tbody>
             {Object.entries(data.marker_genes).map((gene) => (
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/CellInfoSideBar/index.tsx:249]: For readability, let's change `gene` to `[symbol, metadata]` or something better!

File: frontend/src/views/WheresMyGene/components/CellInfoSideBar/style.ts [hunk 23]
```diff
@@ -127,3 +127,8 @@ export const NoMarkerGenesDescription = styled("span")`
     `;
   }}
 `;
+
+export const InfoButtonWrapper = styled("span")`
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/CellInfoSideBar/style.ts:131]: this seems to be the same as `frontend/src/views/WheresMyGene/components/HeatMap/components/YAxisChart/style.ts:79`, maybe worth extracting out as common component in `frontend/src/views/WheresMyGene/components/common/style.ts` or something?

File: frontend/src/views/WheresMyGene/components/GeneInfoSideBar/index.tsx [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+function GeneInfoSideBar({
+  geneInfoGene,
+}: GeneInfoBarProps): JSX.Element | null {
+  const { data, isLoading } = useGeneInfo(geneInfoGene);
+
+  if (isLoading) return <GeneSummary>Loading...</GeneSummary>;
+
+  return (
+    <div id="gene-info-wrapper">
+      {!data ? (
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/GeneInfoSideBar/index.tsx:27]: giant ternary is pretty hard to read, especially with the negative statement first. Maybe we can extract the components like so 🙏 ?

E.g.,

```tsx
import React from "react";
import {
  GeneName,
  GeneSummary,
  GeneSynonyms,
  GeneSynonymsLabel,
  GeneSynonymsWrapper,
  GeneUrl,
  StyledCallout,
} from "./style";
import { useGeneInfo } from "src/common/queries/wheresMyGene";
import { RightSidebarProperties } from "../RightSideBar";
import { GeneInfo } from "../../common/types";

export interface GeneInfoBarProps extends RightSidebarProperties {
  geneInfoGene: string;
}

function GeneInfoSideBar({
  geneInfoGene,
}: GeneInfoBarProps): JSX.Element | null {
  const { data, isLoading } = useGeneInfo(geneInfoGene);

  if (isLoading) return <GeneSummary>Loading...</GeneSummary>;

  const NoData = (
    <>
      <StyledCallout autoDismiss={false} intent={"error"}>
        Sorry, this gene could not be found on NCBI.
      </StyledCallout>

      <GeneUrl
        href={`https://www.google.com/search?q=${geneInfoGene}%20gene`}
        target="_blank"
        rel="noreferrer noopener"
      >
        Search on Google
      </GeneUrl>
    </>
  );

  return (
    <div id="gene-info-wrapper" data-test-id={`${geneInfoGene}-gene-info`}>
      {data ? <Result data={data} /> : NoData}
    </div>
  );
}

function Result({ data }: { data: GeneInfo }) {
  return (
    <>
      <GeneName data-test-id="gene-info-header">{data.name}</GeneName>

      {data.show_warning_banner && (
        <>
          <StyledCallout autoDismiss={false} intent={"warning"}>
            NCBI didn&apos;t return an exact match for this gene.
          </StyledCallout>
        </>
      )}

      <GeneSummary>{data.summary}</GeneSummary>

      <GeneSynonymsWrapper>
        <GeneSynonymsLabel>Synonyms</GeneSynonymsLabel>
        <GeneSynonyms>{data.synonyms.join(", ")}</GeneSynonyms>
      </GeneSynonymsWrapper>

      <GeneUrl href={data.ncbi_url} target="_blank" rel="noreferrer noopener">
        View on NCBI
      </GeneUrl>
    </>
  );
}

export default React.memo(GeneInfoSideBar);

```

File: frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx [hunk 25]
```diff
@@ -66,14 +75,41 @@ function GeneButton({
...
+            data-test-id="gene-info-button-x-axis"
+            className={EXCLUDE_IN_SCREENSHOT_CLASS_NAME}
+            onClick={() => {
+              generateGeneInfo(geneName);
+
+              track(EVENTS.WMG_GENE_INFO, {
+                gene: geneName,
+              });
+
+              if (!dispatch) return;
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx:88]: feels like this guard clause should be before `generateGeneInfo()`, because without dispatch sending analytics and generating gene info don't do anything anyway?

**Comment by signechambers1** [comment]
No, we don't need to track who is clicking on the NCBI link! I can't think of a use case where the tissue matters for gene info off the top of my head so I'm fine with just tracking the gene name for now.

File: frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx (hunk 20 above)
Comment (tihuan) [reply frontend/src/views/WheresMyGene/components/HeatMap/components/XAxisChart/index.tsx:91]: discussion:

I wonder if we could centralize store mutation in reducer instead of burdening event handler to know the store mutation details, effectively following the best practice to use Redux pattern as an event store instead of a command bus

Concretely:

1. Crate an action "selectGenInfoFromXAxis"
2. Reducer gets `selectGenInfoFromXAxis` and add `geneInfo` and remove `cellInfoCellType`

This way the event handler only needs to dispatch `selectGenInfoFromXAxis` and doesn't need to care about what store mutation has to be done!

source:
https://engineering.tableau.com/redux-command-bus-or-event-store-2c4c044cd481#ab5e

**Review by tihuan — COMMENTED** [review]
Feature works really well! I also checked SVG download and it seems unaffected too 👍 

Part 1 of review, since this is a huge PR 😆

File: frontend/src/common/queries/wheresMyGene.ts [hunk 26]
```diff
@@ -1219,6 +1252,30 @@ export function useMarkerGenes({
...
+
+    result = generateTermsByKey(genes, "name");
+
+    return result;
+  }, [data]);
+
+  const geneID = genesByName[geneSymbol]?.id;
+  return useQuery(
+    [USE_GENE_INFO, geneSymbol, geneID],
+    ({ signal }) => fetchGeneInfo({ geneSymbol, geneID, signal }),
```
[the remark this answers is not in the document]
Comment (atarashansky) [reply frontend/src/common/queries/wheresMyGene.ts:1272]: https://github.com/chanzuckerberg/single-cell-data-portal/blob/6ecc8f64c80287d28da853dabb373a7a6ae46830/backend/gene_info/api/ncbi_provider.py#L49

>>> Given a gene ensembl ID and gene name, returns a tuple with NCBI's corresponding gene UID and a boolean noting if the result is from searching by the gene name, instead of the gene ENSEMBL id. Initially, uses ensembl ID to find gene UID, but in the event that this returns an unexpected result, call the NCBI API again with gene name. This is successful if the response returns only 1 result for UID.

TL;DR - sometimes using the UUID doesn't work if it's out-of-date and so the gene symbol is a fallback.

File: frontend/tests/features/wheresMyGene.test.ts [hunk 27]
```diff
@@ -1020,6 +1122,39 @@ async function waitForElement(page: Page, selector: string) {
   );
 }
 
+async function waitForElementToBeRemoved(page: Page, selector: string) {
```
Comment (tihuan) [inline frontend/tests/features/wheresMyGene.test.ts:1125]: Great utility function! Maybe move this to  `frontend/tests/utils/helpers.ts` ?

**Review by tihuan — APPROVED** [review]
Awesome work, Andrew 🎉 !  Really appreciate the great quality PR too 🙏  

Just some minor nits left otherwise LGTM 🔥
